pkg/overlay/server.go

```diff
@@ -37,59 +33,100 @@ func (o *Server) Lookup(ctx context.Context, req *proto.LookupRequest) (*proto.L
 	}
 
 	return &proto.LookupResponse{
-		Node: &proto.Node{
-			Id:      req.GetNodeID(),
-			Address: na,
-		},
+		Node: na,
 	}, nil
 }
 
 // FindStorageNodes searches the overlay network for nodes that meet the provided requirements
-func (o *Server) FindStorageNodes(ctx context.Context, req *proto.FindStorageNodesRequest) (*proto.FindStorageNodesResponse, error) {
-	// NB:  call FilterNodeReputation from node_reputation package to find nodes for storage
-	keys, err := o.cache.DB.List(nil, storage.Limit(10))
+func (o *Server) FindStorageNodes(ctx context.Context, req *proto.FindStorageNodesRequest) (resp *proto.FindStorageNodesResponse, err error) {
+	opts := req.GetOpts()
+	maxNodes := opts.GetLimit()
+	restrictions := opts.GetRestrictions()
+	restrictedBandwidth := restrictions.GetFreeBandwidth()
+	restrictedSpace := restrictions.GetFreeDisk()
+
+	var start storage.Key
+	result := []*proto.Node{}
+	for {
+		var nodes []*proto.Node
+		nodes, start, err = o.populate(ctx, start, maxNodes, restrictedBandwidth, restrictedSpace)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(nodes) <= 0 {
+			break
+		}
+
+		result = append(result, nodes...)
+
+		if len(result) >= int(maxNodes) || start == nil {
+			break
+		}
+
+	}
+
+	if len(result) > int(maxNodes) {
+		result = result[:maxNodes]
+	}
+
+	return &proto.FindStorageNodesResponse{
+		Nodes: result,
+	}, nil
+}
+
+func (o *Server) getNodes(ctx context.Context, keys storage.Keys) ([]*proto.Node, error) {
+	values, err := o.cache.DB.GetAll(keys)
 	if err != nil {
-		o.logger.Error("Error listing nodes", zap.Error(err))
 		return nil, err
 	}
 
-	if len(keys) > maxNodes {
-		// TODO(coyle): determine if this is a set value or they user of the api will specify
-		keys = keys[:maxNodes]
+	nodes := []*proto.Node{}
+	for _, v := range values {
+		n := &proto.Node{}
+		if err := protob.Unmarshal(v, n); err != nil {
+			return nil, err
+		}
+
+		nodes = append(nodes, n)
 	}
 
-	nodes := o.getNodes(ctx, keys)
+	return nodes, nil
 
-	return &proto.FindStorageNodesResponse{
-		Nodes: nodes,
-	}, nil
 }
 
-func (o *Server) getNodes(ctx context.Context, keys storage.Keys) []*proto.Node {
-	wg := &sync.WaitGroup{}
-	ch := make(chan *proto.Node, len(keys))
+func (o *Server) populate(ctx context.Context, starting storage.Key, maxNodes, restrictedBandwidth, restrictedSpace int64) ([]*proto.Node, storage.Key, error) {
+	limit := storage.Limit(maxNodes) * 2
+	keys, err := o.cache.DB.List(nil, limit)
```

Comment:
doesn't look like we use the starting key do we?

Reply:
great catch